neuralmonkey/encoders/sentence_cnn_encoder.py

```diff
@@ -0,0 +1,278 @@
+                    w_filter = tf.get_variable(
+                        "conv_W", filter_shape,
+                        initializer=tf.random_uniform_initializer(-0.5, 0.5))
+                    b_filter = tf.get_variable(
+                        "conv_bias", [num_filters],
+                        initializer=tf.constant_initializer(0.0))
+                    conv = tf.nn.conv1d(
+                        embedded_inputs,
+                        w_filter,
+                        stride=1,
+                        padding="VALID",
+                        name="conv")
+
+                    # Apply nonlinearity
+                    conv_relu = tf.nn.relu(tf.nn.bias_add(conv, b_filter))
+
+                    # Max-pooling over the output segments
+                    expanded_conv_relu = tf.expand_dims(conv_relu, -1)
+                    pooled = tf.nn.max_pool(
+                        expanded_conv_relu,
+                        ksize=[1, self.segment_size, 1, 1],
+                        strides=[1, self.segment_size, 1, 1],
+                        padding="SAME")
+                    pooled_outputs.append(pooled)
+
+            # Combine all the pooled features
+            self.cnn_encoded = tf.concat(pooled_outputs, axis=2)
+            self.cnn_encoded = tf.squeeze(self.cnn_encoded, [3])
+
+            # Highway Network
+            batch_size = tf.shape(self.cnn_encoded)[0]
+            # pylint: disable=no-member
+            cnn_out_size = self.cnn_encoded.get_shape().as_list()[-1]
+            highway_layer = tf.reshape(self.cnn_encoded, [-1, cnn_out_size])
+            for i in range(self.highway_depth):
+                highway_layer = highway(
+                    highway_layer,
+                    scope=("highway_layer_%s" % i))
+            highway_layer = tf.reshape(
+                highway_layer,
+                [batch_size, -1, cnn_out_size])
+
+            # BiRNN Network
+            fw_cell, bw_cell = self.rnn_cells()  # type: RNNCellTuple
+            seq_lens = tf.ceil(tf.divide(
+                self.sentence_lengths,
+                self.segment_size))
+            seq_lens = tf.cast(seq_lens, tf.int32)
+            outputs_bidi_tup, encoded_tup = tf.nn.bidirectional_dynamic_rnn(
+                fw_cell, bw_cell, highway_layer,
+                sequence_length=seq_lens,
+                dtype=tf.float32)
+
+            self.hidden_states = tf.concat(outputs_bidi_tup, 2)
+
+            with tf.variable_scope('attention_tensor'):
+                self.__attention_tensor = self._dropout(
+                    self.hidden_states)
+
+            self.encoded = tf.concat(encoded_tup, 1)
+
+        log("Sentence encoder initialized")
+
+    @property
+    def _attention_tensor(self):
+        return self.__attention_tensor
+
+    @property
+    def _attention_mask(self):
+        # TODO tohle je proti OOP prirode
+        return self.input_mask
```

Comment:
Ta maska by se měla spočítat ze `seq_lens` pomocí `tf.sequence_mask`, ne? Takhle to v dekodéru spadne.